pkg/store/proxy.go

```diff
@@ -49,12 +63,37 @@ type ProxyStore struct {
 	selectorLabels labels.Labels
 
 	responseTimeout time.Duration
+	metrics         *proxyStoreMetrics
+}
+
+func newProxyStoreMetrics(reg *prometheus.Registry) *proxyStoreMetrics {
+	var m proxyStoreMetrics
+
+	m.firstRecvDuration = prometheus.NewSummaryVec(prometheus.SummaryOpts{
+		Name:       "thanos_proxy_first_recv_duration",
+		Help:       "Time to get first part data from store(ms).",
+		Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		MaxAge:     2 * time.Minute,
+	}, []string{"store", "payload"})
```

Comment:
We talked about `store` in metrics - this might leak cardinaltity (changing IP address), so I think we have to hook it to external labels and store TYPE as we do in storeset.

Reply:
We have too large external labels in our thanos-store's, therefore external labels is not comfortable. Also external_labels can have different order time from time. Maybe we can think about comfortable way to identify stores? For example, by bucket_name?

Reply:
See storeset, we already do that (we also sort it), so if that's the problem we need to fix it everywhere (:

Reply:
![Снимок экрана 2019-11-27 в 17 58 38](https://user-images.githubusercontent.com/2960110/69734170-97772080-113f-11ea-8057-09f1a3eb5f29.png)
I can't imagine comfortable work with such external_labels, presented as json string

Reply:
Can you show querier /metrics page?

In this case, we need to fix this in separate PR `stores hash as you already have metric with this label: https://github.com/thanos-io/thanos/blob/7e11afe64af0c096743a3de8a594616abf52be45/pkg/query/storeset.go#L94 (:

Reply:
`thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-0\",prometheus_replica=\"A\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-0\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-0\",prometheus_replica=\"B\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-1\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-1\",prometheus_replica=\"B\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-2\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-2\",prometheus_replica=\"B\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-3\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-3\",prometheus_replica=\"B\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-4\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-4\",prometheus_replica=\"B\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-5\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-5\",prometheus_replica=\"B\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-0\",prometheus_replica=\"B\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-1\",prometheus_replica=\"A\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-1\",prometheus_replica=\"B\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-2\",prometheus_replica=\"A\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-2\",prometheus_replica=\"B\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-3\",prometheus_replica=\"A\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-3\",prometheus_replica=\"B\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-4\",prometheus_replica=\"A\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-4\",prometheus_replica=\"B\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-5\",prometheus_replica=\"A\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-5\",prometheus_replica=\"B\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-hardware-0\",prometheus_replica=\"A\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-hardware-0\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-hardware-0\",prometheus_replica=\"B\"},{dc=\"mts\",env=\"production\",prometheus_replica=\"prometheus01z1.h.o3.ru\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-hardware-0\",prometheus_replica=\"B\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-ingress-0\",prometheus_replica=\"A\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-ingress-0\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-ingress-0\",prometheus_replica=\"B\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-ingress-1\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-ingress-1\",prometheus_replica=\"B\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-other-0\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-other-0\",prometheus_replica=\"B\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-system-0\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-system-0\",prometheus_replica=\"B\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-ingress-0\",prometheus_replica=\"B\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-ingress-1\",prometheus_replica=\"A\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-ingress-1\",prometheus_replica=\"B\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-other-0\",prometheus_replica=\"A\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-other-0\",prometheus_replica=\"B\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-system-0\",prometheus_replica=\"A\"}"} 1
thanos_store_node_info{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-system-0\",prometheus_replica=\"B\"}"} 1
# HELP thanos_store_nodes_grpc_connections Number of gRPC connection to Store APIs. Opened connection means healthy store APIs available for Querier.
# TYPE thanos_store_nodes_grpc_connections gauge
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-0\",prometheus_replica=\"A\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-0\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-0\",prometheus_replica=\"B\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-1\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-1\",prometheus_replica=\"B\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-2\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-2\",prometheus_replica=\"B\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-3\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-3\",prometheus_replica=\"B\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-4\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-4\",prometheus_replica=\"B\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-5\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-5\",prometheus_replica=\"B\"}",store_type="store"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-0\",prometheus_replica=\"B\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-1\",prometheus_replica=\"A\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-1\",prometheus_replica=\"B\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-2\",prometheus_replica=\"A\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-2\",prometheus_replica=\"B\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-3\",prometheus_replica=\"A\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-3\",prometheus_replica=\"B\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-4\",prometheus_replica=\"A\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-4\",prometheus_replica=\"B\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-5\",prometheus_replica=\"A\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-apps-5\",prometheus_replica=\"B\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-hardware-0\",prometheus_replica=\"A\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-hardware-0\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-hardware-0\",prometheus_replica=\"B\"},{dc=\"mts\",env=\"production\",prometheus_replica=\"prometheus01z1.h.o3.ru\"}",store_type="store"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-hardware-0\",prometheus_replica=\"B\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-ingress-0\",prometheus_replica=\"A\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-ingress-0\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-ingress-0\",prometheus_replica=\"B\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-ingress-1\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-ingress-1\",prometheus_replica=\"B\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-other-0\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-other-0\",prometheus_replica=\"B\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-system-0\",prometheus_replica=\"A\"},{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-system-0\",prometheus_replica=\"B\"}",store_type="store"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-ingress-0\",prometheus_replica=\"B\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-ingress-1\",prometheus_replica=\"A\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-ingress-1\",prometheus_replica=\"B\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-other-0\",prometheus_replica=\"A\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-other-0\",prometheus_replica=\"B\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-system-0\",prometheus_replica=\"A\"}",store_type="sidecar"} 1
thanos_store_nodes_grpc_connections{external_labels="{dc=\"m9\",env=\"production\",prometheus=\"monitoring/prometheus-system-0\",prometheus_replica=\"B\"}",store_type="sidecar"} 1`
All metrics with `external_labels`. It looks ugly and unreadable, also json string to hard to process in grafana (for naming graphs)

Reply:
It's crazy!

I think, still this should be resolved outside of this PR (: 

Again, I think `address` might add some cardinality, but I am tempted to allow it under some flag...